pkg/vault/contracts/Vault.sol

```diff
@@ -1148,44 +1146,55 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
+                    vaultWrappedDelta = wrappedToken.previewDeposit(vaultUnderlyingDelta);
+                } else {
+                    underlyingToken.forceApprove(address(wrappedToken), vaultUnderlyingDelta);
+                    // EXACT_IN requires the exact amount of underlying tokens to be deposited, so deposit is called.
+                    vaultWrappedDelta = wrappedToken.deposit(vaultUnderlyingDelta, address(this));
+                }
             } else {
-                // Note that `bufferWrappedSurplus` will be zero if there is no bufferUnderlyingSurplus.
-                calculatedWrappedDelta = amountOutWrapped + bufferWrappedSurplus;
+                if (bufferUnderlyingSurplus > 0) {
+                    bufferWrappedSurplus = wrappedToken.convertToShares(bufferUnderlyingSurplus);
+                }
+                vaultWrappedDelta = amountOutWrapped + bufferWrappedSurplus;
+
+                if (
+                    isQueryContext &&
+                    underlyingToken.balanceOf(address(this)) < amountInUnderlying + bufferUnderlyingSurplus
+                ) {
+                    vaultUnderlyingDelta = wrappedToken.previewMint(vaultWrappedDelta);
+                } else {
+                    // Add convert error because mint can consume a different amount of tokens than we anticipated.
+                    underlyingToken.forceApprove(
+                        address(wrappedToken),
+                        _addConvertError(amountInUnderlying + bufferUnderlyingSurplus)
+                    );
+
+                    // EXACT_OUT requires the exact amount of wrapped tokens to be returned, so mint is called.
+                    vaultUnderlyingDelta = wrappedToken.mint(vaultWrappedDelta, address(this));
+
+                    // Remove approval, in case mint consumed less tokens than we approved, due to convert error.
+                    underlyingToken.forceApprove(address(wrappedToken), 0);
+                }
+            }
 
-                // Add convert error because mint can consume a different amount of tokens than we anticipated.
-                underlyingToken.forceApprove(
-                    address(wrappedToken),
-                    _addConvertError(amountInUnderlying + bufferUnderlyingSurplus)
+            if (isQueryContext == false) {
+                // Do not trust ERC4626 and measure the amount of deposited and returned tokens.
+                (vaultUnderlyingDelta, vaultWrappedDelta) = _updateReservesAfterWrapping(
+                    underlyingToken,
+                    IERC20(wrappedToken)
                 );
-
-                // EXACT_OUT requires the exact amount of wrapped tokens to be returned, so mint is called.
-                wrappedToken.mint(calculatedWrappedDelta, address(this));
-
-                // Remove approval, in case mint consumed less tokens than we approved, due to convert error.
-                underlyingToken.forceApprove(address(wrappedToken), 0);
             }
```

Comment:
If we are in a query context, can't we update the reserves anyways?

Reply:
ah, but there's no real operation in case we preview. Maybe I'd explain why in a query context we can't always do this, and why it doesn't matter?

Reply:
Done!